language/move-symbol-pool/src/symbol.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{pool::Entry, SYMBOL_POOL};
+use serde::{Deserialize, Serialize};
+use std::{cmp::Ordering, fmt, num::NonZeroU64};
+
+/// Represents a string that has been cached.
+///
+/// A `Symbol` represents a pointer to string data that is owned by the global
+/// symbol pool; it is not the string data itself. This enables this
+/// representation to implement `Copy` and other traits that some string types
+/// cannot.
+///
+/// The strings that `Symbol` types represent are added to the global cache as
+/// the `Symbol` are created.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let s1 = Symbol::from("hi"); // "hi" is stored in the global cache
+/// let s2 = Symbol::from("hi"); // "hi" is already stored, cache does not grow
+/// assert_eq!(s1, s2);
+/// ```
+///
+/// Use the method [`as_str()`] to access the string value that a `Symbol`
+/// represents. `Symbol` also implements the [`Display`] trait, so it can be
+/// printed as an ordinary string would. This makes it easier to use with
+/// crates that print strings to a terminal, such as codespan.
+///
+/// ```
+///# use crate::move_symbol_pool::Symbol;
+/// let message = format!("{} {}",
+///     Symbol::from("hello").as_str(),
+///     Symbol::from("world"));
+/// assert_eq!(message, "hello world");
+/// ```
+///
+/// [`as_str()`]: crate::move_symbol_pool::Symbol::as_str
+/// [`Display`]: std::fmt::Display
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct Symbol(NonZeroU64);
+
+impl Symbol {
+    pub fn as_str<'a>(&self) -> &'a str {
```

Comment:
wild... 

If this is true, should we just do `&'static str` Or are there lifetimes that aren't less than static?

Reply:
Might make more sense to bind the lifetime to `&self` that way the caller can't just arbitrarily choose the lifetime they want
```suggestion
    pub fn as_str(&self) -> &str {
```

Reply:
I went with a `&self`-bound lifetime over a static one, but both type-check successfully. I don't know enough about Rust to have an opinion here. Since symbols are never removed from the pool, their underlying strings and references to them truly are static, and so AFAICT not necessarily bound to the lifetime of the `Symbol`.

Reply:
they are technically static as long as we don't drop anything.
Binding it to self is the best longterm decision if we ever do the symbol inlining thing